src/core_plugins/vega/public/vega_view/vega_base_view.js

```diff
@@ -0,0 +1,176 @@
+        };
+        const loader = vega.loader();
+        loader.load = errorFunc;
+        loader.sanitize = errorFunc;
+        this._vegaViewConfig.loader = loader;
+      }
+
+      // The derived class should create this method
+      await this._initViewCustomizations();
+    } catch (err) {
+      this.onError(err);
+    }
+  }
+
+  onError() {
+    this._addMessage('err', Utils.formatErrorToStr(...arguments));
+  }
+
+  onWarn() {
+    if (!this._parser || !this._parser.hideWarnings) {
+      this._addMessage('warn', Utils.formatWarningToStr(...arguments));
+    }
+  }
+
+  _addMessage(type, text) {
+    if (!this._$messages) {
+      this._$messages = $(`<ul class="vega-messages">`).appendTo(this._$parentEl);
+    }
+    this._$messages.append(
+      $(`<li class="vega-message-${type}">`).append(
+        $(`<pre>`).text(text)
+      )
+    );
+  }
+
+  resize() {
+    if (this._parser.useResize && this._view && this.updateVegaSize(this._view)) {
+      return this._view.runAsync();
+    }
+  }
+
+  updateVegaSize(view) {
+    // For some reason the object is slightly scrollable without the extra padding.
+    // This might be due to https://github.com/jquery/jquery/issues/3808
+    // Which is being fixed as part of jQuery 3.3.0
+    const heightExtraPadding = 6;
+    const width = Math.max(0, this._$container.width() - this._parser.paddingWidth);
+    const height = Math.max(0, this._$container.height() - this._parser.paddingHeight) - heightExtraPadding;
+    if (view.width() !== width || view.height() !== height) {
+      view.width(width).height(height);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Set global debug variable to simplify vega debugging in console. Show info message first time
+   */
+  setDebugValues(view, spec, vlspec) {
+    if (!this._editorMode) {
+      // VEGA_DEBUG should only be enabled in the editor mode
+      return;
+    }
+
+    if (window) {
+      if (!view) {
+        // disposing, get rid of the stale debug info
+        delete window.VEGA_DEBUG;
+      } else {
+        if (window.VEGA_DEBUG === undefined && console) {
+          console.log('%cWelcome to Kibana Vega Plugin!', 'font-size: 16px; font-weight: bold;');
+          console.log('You can access the Vega view with VEGA_DEBUG. ' +
+            'Learn more at https://vega.github.io/vega/docs/api/debugging/.');
+        }
+
+        window.VEGA_DEBUG = window.VEGA_DEBUG || {};
+        window.VEGA_DEBUG.VEGA_VERSION = vega.version;
+        window.VEGA_DEBUG.VEGA_LITE_VERSION = vegaLite.version;
+        window.VEGA_DEBUG.view = view;
+        window.VEGA_DEBUG.vega_spec = spec;
+        window.VEGA_DEBUG.vegalite_spec = vlspec;
+      }
+    }
+  }
+
+  destroy() {
+    // properly handle multiple destroy() calls by converting this._destroyHandlers
+    // from an array into a promise, while handlers are being disposed
+    if (this._destroyHandlers) {
+      if (this._destroyHandlers.then) {
```

Comment:
I guess I wouldn't use the same variable, to store the ongoing process of destroying and the previously registered destroy listeners. I think it makes this less readable, than something like follows:

```js
destroy() {
  if (this._destroyHandlers) {
    if (!this._ongoingDestroy) {
      this._ongoingDestroy = Promise.all(this._destroyHandlers.map(v => v())
        .then(() => {
          this._destroyHandlers = null;
          this._ongoingDestroy = null;
        });
    }
    return this._ongoingDestroy;
  }
}

_addDestroyHandler(handler) {
  if (this._ongoingPromise) {
    handler();
  } else {
    this._destroyHandlers.push(handler);
  }
}
```

I btw still think that code has an issue. In both ways, when the destroy is already ongoing and we call `_addDestroyHandler`, we execute the handler immediately, but we don't actual wait for it to finish, since the `Promise.all` will always resolve once the initial destroy handlers are finished, ignoring those later once.